packages/tpetra/core/src/Tpetra_Import_def.hpp

```diff
@@ -841,23 +843,32 @@ namespace Tpetra {
 
     // The input map is *not* necessarily sorted, but the "same" GIDs
     // are still guaranteed to be at the beginning.  Copy the same GIDs to the
-    // unionTgtGIDs vector.
+    // unionTgtGIDs vector.  Cases for numSameGIDs1 != numSameGIDs2 must be
+    // treated separately.
     if (numSameGIDs1 > numSameGIDs2) {
+
       numSameGIDs = numSameGIDs1;
-      std::copy(tgtGIDs1.begin(), tgtGIDs1.begin()+numSameGIDs, std::back_inserter(unionTgtGIDs));
+
+      // Copy the same GIDs from tgtGIDs to the union
+      std::copy(tgtGIDs1.begin(), tgtGIDs1.begin()+numSameGIDs1,
+                std::back_inserter(unionTgtGIDs));
+
       // Shift GIDs in notSame2 already in unionTgtGIDs to end and remove them.
-      // See comments below about usage of stable_partition
       auto it = std::stable_partition(notSame2.begin(), notSame2.end(),
         [&](GO GID){
           return !std::binary_search(unionTgtGIDs.begin(), unionTgtGIDs.end(), GID);
         }
       );
       notSame2.erase(it, notSame2.end());
```

Comment:
Instead of erasing, the iterator can be saved and used instead of `notSame2.end()` in any code that follows.